### PR TITLE
extras: add covariance parsing to vision_speed_estimate

### PR DESCRIPTION
**Pull request by TSC21**

### CODE REVIEW

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 1]
```diff
@@ -46,6 +48,7 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
 
 		if (listen_twist)
 			vision_vel_sub = sp_nh.subscribe("speed_twist", 10, &VisionSpeedEstimatePlugin::vel_twist_cb, this);
+
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: Should not be there.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 2]
```diff
@@ -62,12 +65,11 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
 
 	/* -*- low-level send -*- */
 
-	void vision_speed_estimate(uint64_t usec, Eigen::Vector3d &v)
+	void vision_speed_estimate(uint64_t usec, const Eigen::Vector3d &v, const mavros::ftf::Covariance3d &cov)
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: `mavros` namespace shold be already used. So just `ftf::Covariance3d`.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 3]
```diff
@@ -76,38 +78,53 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
 		vs.y = v.y();
 		vs.z = v.z();
 		// [[[end]]] (checksum: aee3cc9a73a2e736b7bc6c83ea93abdb)
+		ftf::covariance_to_mavlink(cov, vs.covariance);
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: Leave blank lines before and after cog block.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 4]
```diff
@@ -76,38 +78,53 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
...
-		vision_speed_estimate(stamp.toNSec() / 1000, vel);
+		vision_speed_estimate(stamp.toNSec() / 1000, vel_ned, cov_ned);
 	}
 
 	/* -*- callbacks -*- */
 
-	void vel_twist_cb(const geometry_msgs::TwistStamped::ConstPtr &req) {
-		send_vision_speed(req->twist.linear, req->header.stamp);
+	void vel_twist_with_cov_cb(const geometry_msgs::TwistStamped::ConstPtr &req) {
+		ftf::Covariance3d cov {};	// zero initialized
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: Does mavlink propose some value for incorrect cov? Like [-1, 0, 0...] in ROS msgs.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 5]
```diff
@@ -76,38 +78,53 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
...
+	void vel_twist_cb(const geometry_msgs::TwistWithCovarianceStamped::ConstPtr &req) {
+		ftf::Covariance3d cov3d {};	// zero initialized
+		ftf::Covariance6d cov6d {};	// zero initialized
+
+		ftf::EigenMapCovariance3d cov3d_map(cov3d.data());
+		ftf::EigenMapCovariance6d cov6d_map(cov6d.data());
+
+		// only the linear velocity will be sent
+		cov6d = req->twist.covariance;
+		cov3d_map = cov6d_map.block<3, 3>(0, 0);
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: Unnecessary copy.

```cpp
ftf::EigenMapConstCovariance6d cov6d_map(req->twist.covariance);

cov3d_map = cov6d_map.block<3, 3>(0, 0);
```

File: mavros_extras/src/plugins/vision_speed_estimate.cpp (hunk 1 above)
Comment (TSC21) [reply mavros_extras/src/plugins/vision_speed_estimate.cpp]: actually I did this PR quickly and didn't realize I forgot to add the `speed_twist_cov` option.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp (hunk 4 above)
Comment (TSC21) [reply mavros_extras/src/plugins/vision_speed_estimate.cpp]: It doesn't. But should be a good idea to bring yes.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp (hunk 3 above)
Comment (TSC21) [reply mavros_extras/src/plugins/vision_speed_estimate.cpp]: Any good reason for it or is just for aesthetics?

File: mavros_extras/src/plugins/vision_speed_estimate.cpp (hunk 5 above)
Comment (TSC21) [reply mavros_extras/src/plugins/vision_speed_estimate.cpp]: Makes sense. Thanks

Comment (TSC21) [reply mavros_extras/src/plugins/vision_speed_estimate.cpp]: `ftf::EigenMapConstCovariance6d cov6d_map(req->twist.covariance.data());` btw.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 6]
```diff
@@ -58,16 +64,20 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
...
 
 	/* -*- low-level send -*- */
 
-	void vision_speed_estimate(uint64_t usec, Eigen::Vector3d &v)
+	void vision_speed_estimate(uint64_t usec, const Eigen::Vector3d &v, const ftf::Covariance3d &cov)
 	{
 		mavlink::common::msg::VISION_SPEED_ESTIMATE vs{};
 
 		vs.usec = usec;
-
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: You forgot to remove that change.

File: mavros_extras/src/plugins/vision_speed_estimate.cpp [hunk 7]
```diff
@@ -77,37 +87,51 @@ class VisionSpeedEstimatePlugin : public plugin::PluginBase {
...
 	 */
-	void send_vision_speed(const geometry_msgs::Vector3 &vel_enu, const ros::Time &stamp)
+	void send_vision_speed(const ros::Time &stamp, const geometry_msgs::Vector3 &vel_enu, const ftf::Covariance3d &cov)
 	{
-		Eigen::Vector3d vel_;
-		tf::vectorMsgToEigen(vel_enu, vel_);
-		//Transform from ENU to NED frame
-		auto vel = ftf::transform_frame_enu_ned(vel_);
+		Eigen::Vector3d vel;
+		tf::vectorMsgToEigen(vel_enu, vel);
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_speed_estimate.cpp]: Use `ftf::to_eigen()`, lets remove most of `tf:vectorMsgToEigen()`.

**Comment by TSC21** [comment]
> What about covariance unused/unknown special value? All 0?

Still no special value. I will do a RFC on Mavlink to that respect.

**Comment by TSC21** [comment]
Jenkins test this please.

**Comment by vooon** [comment]
Jenkins already test it. Builds ok, just mavconn test segfaults again. Also i'm not sure that ros jenkins watch on comments.

When you have `if` inside `if` block it is safer to explicitly define block.

**Pull request merged**